### PR TITLE
Divisibility by 32

### PR DESCRIPTION
**Pull request by II245**

### CODE REVIEW

File: tools.py [hunk 1]
```diff
@@ -509,17 +510,62 @@ def split(
...
+
+	win_length = int(window_size * sample_rate)
+	hop_length = int(window_stride * sample_rate)
+	nfft = 2 ** math.ceil(math.log2(win_length))
+
+	freq_cutoff = nfft // 2 + 1
+	additional_padding = freq_cutoff - 1  # additional_padding uses in fronted, two times for mirror and constant pad
+
+	# used formula from https://pytorch.org/docs/stable/generated/torch.nn.Conv1d.html
+	def get_out_shape(l_in):
```
Comment (vadimkantorov) [inline tools.py]: please move the shape computation to a method in LogFilterBankFrontend class, they are tightly coupled and should be placed next to each other

File: tools.py [hunk 2]
```diff
@@ -640,9 +686,18 @@ def split(
 	cmd.add_argument('--comment-path', '-c')
 	cmd.add_argument('--map-tag', action = type('', (argparse.Action, ), dict(__call__ = lambda a, p, n, v, o: getattr(n, a.dest).update(dict([v.split('=')])))), default = dict())
 	cmd.add_argument('--stop-tag', action = 'append', default = [])
-	
 	cmd.set_defaults(func = wordtags)
 
+	cmd = subparsers.add_parser('find_shapes')
```
Comment (vadimkantorov) [inline tools.py]: please synchronize the method name and the command name

File: tools.py [hunk 3]
```diff
@@ -509,17 +510,62 @@ def split(
...
+
+'''
+This script helps to find solution for input and output shapes of signal with frontend divisibility restrictions
+
+example:
+python tools.py find_shapes --start 119 --end 121 --input-shape-restriction 16 --output-shape-restriction 32
+'''
+
+
+def find_solution_for_frontend_input_output_shapes_divisibility(
```
Comment (vadimkantorov) [inline tools.py:533]: after making the proposed edits, please submit this functionality in a separate PR, so that we can merge this without modifying the frontend padding logic

File: tools.py (hunk 2 above)
Comment (II245) [reply tools.py]: done

File: tools.py (hunk 1 above)
Comment (II245) [reply tools.py]: done

File: tools.py (hunk 3 above)
Comment (II245) [reply tools.py:533]: done

File: models.py [hunk 4]
```diff
@@ -584,8 +607,27 @@ def forward(self, signal: shaping.BT, mask: shaping.BT = None, **kwargs) -> shap
...
+		if self.check_features_divisibility_by_32:
+			if signal.shape[-1] % (32 / 2) != 0:
+				warnings.warn('Shape of input signal not divisible by 16')
+
+			if log_mel_features.shape[-1] % 32 != 0:
+				warnings.warn('Shape of frontend output signal not divisible by 32')
+
 		return log_mel_features
 
+	@staticmethod
```
Comment (vadimkantorov) [inline models.py:591]: please move this method in the other PR. the idea is to merge in just the output shape optimization, without modified padding behavior

File: models.py [hunk 5]
```diff
@@ -584,8 +607,27 @@ def forward(self, signal: shaping.BT, mask: shaping.BT = None, **kwargs) -> shap
...
+	# used formula from https://pytorch.org/docs/stable/generated/torch.nn.Conv1d.html
+	def get_out_shape(l_in, kernel_size, stride, additional_padding, padding=0, dilation=1):
```
Comment (vadimkantorov) [inline models.py]: get_out_shape -> compute_output_shape

`l_in` -> something more semantic. `l_in` is uncomprehensible by itself

why do you call it `additional_padding`? we're dealing with regular padding, right?

File: models.py [hunk 6]
```diff
@@ -584,8 +607,27 @@ def forward(self, signal: shaping.BT, mask: shaping.BT = None, **kwargs) -> shap
...
+		# additional_padding uses in fronted, two times for mirror and constant pad
+		return int(math.floor(((l_in + 2 * additional_padding) + 2 * padding - dilation * (kernel_size - 1) - 1) / stride + 1))
```
Comment (vadimkantorov) [inline models.py]: please use `padding` instead of `additional_padding`. existing `padding` term in the Conv1d is exactly what we're dealing with

Comment (II245) [reply models.py]: done

File: models.py (hunk 5 above)
Comment (II245) [reply models.py]: done

File: models.py (hunk 4 above)
Comment (II245) [reply models.py:591]: done

File: benchmark.py [hunk 7]
```diff
@@ -81,10 +84,12 @@
 
 tictoc = lambda: (use_cuda and torch.cuda.synchronize()) or time.time()
 
-batch_shape = [args.B, args.T * args.sample_rate
+batch_shape = [args.B, int(args.T * args.sample_rate)
 				] if args.frontend else [args.B, args.num_input_features, int(args.T / args.window_stride)]
-if batch_shape[-1] % 128 != 0:
+
+if args.satisfy_t_shape_divisibility_by_128 and batch_shape[-1] % 128 != 0:
```
Comment (vadimkantorov) [inline benchmark.py]: please remove constant duplication. instead of `satisfy_t_shape_divisibility_by_128 `, please make it an `int` argparse  argument `input_time_dim_multiple` - just like you did elsewhere (and set default value to 128)

File: models.py [hunk 8]
```diff
@@ -584,8 +587,22 @@ def forward(self, signal: shaping.BT, mask: shaping.BT = None, **kwargs) -> shap
 
 		power_spectrum = real_squared + imag_squared
 		log_mel_features = self.mel(power_spectrum).log()
+
+		if self.check_features_divisibility_by_32:
```
Comment (vadimkantorov) [inline models.py]: -> `check_time_dim_padded`

Comment (vadimkantorov) [inline models.py]: please move this check into JasperNet's forward, let's start with the check there

File: models.py [hunk 9]
```diff
@@ -584,8 +587,22 @@ def forward(self, signal: shaping.BT, mask: shaping.BT = None, **kwargs) -> shap
...
+			if signal.shape[-1] % (32 / 2) != 0:
+				warnings.warn('Shape of input signal not divisible by 16')
```
Comment (vadimkantorov) [inline models.py]: please use  `assert condition, 'message'` like we're doing elsewhere. assert is sufficient for our usecase

File: benchmark.py [hunk 10]
```diff
@@ -30,12 +30,14 @@
 parser.add_argument('--onnx')
 parser.add_argument('--stft-mode', choices = ['conv', ''], default = '')
 parser.add_argument('-B', type = int, default = 256)
-parser.add_argument('-T', type = int, default = 5.12)
+parser.add_argument('-T', type = float, default = 5.12)
```
Comment (vadimkantorov) [inline benchmark.py:33]: Why is non-integer float T needed?

Comment (II245) [reply benchmark.py:33]: For cases when needed to test divisibility. I have T values like 129.99 seconds.

File: models.py (hunk 8 above)
Comment (II245) [reply models.py]: done

Comment (II245) [reply models.py]: done

File: models.py (hunk 9 above)
Comment (II245) [reply models.py]: done

File: benchmark.py (hunk 7 above)
Comment (II245) [reply benchmark.py]: done
